packages/ui/src/components/notifications/NotificationCenter.tsx

```diff
@@ -0,0 +1,88 @@
+import { ComponentType, useEffect, useState } from 'react';
+
+import { Notification } from './types';
+
+const NOTIFICATION_EVENT = 'emit-core-notification';
+const DEFAULT_CACHE_TIMEOUT = 5 * 60 * 1000; // 5 minutes in milliseconds
+const DEFAULT_MAX_NOTIFICATIONS = 5;
+
+export const notify = (props: Omit<Notification, 'id' | 'createdAt'>) => {
+    const id = window.crypto.randomUUID();
+    const createdAt = new Date();
+    const event = new CustomEvent(NOTIFICATION_EVENT, {
+        detail: { id, open: true, createdAt, ...props },
+    });
+    document.dispatchEvent(event);
+};
+
+type SnackbarComponentProps = Notification & {
+    onClose: (id: string) => void;
+};
+
+type NotificationCenterProps = {
+    maxNotifications?: number;
+    hideAfter?: number;
+    SnackbarComponent: ComponentType<SnackbarComponentProps>;
+    cacheTimeout?: number;
+    className?: string;
+};
+
+const NotificationCenter = ({
+    SnackbarComponent,
+    hideAfter = 0,
+    cacheTimeout = DEFAULT_CACHE_TIMEOUT,
+    maxNotifications = DEFAULT_MAX_NOTIFICATIONS,
+    className,
+}: NotificationCenterProps) => {
+    const [notifications, setNotifications] = useState<Notification[]>([]);
+
+    const handleCloseNotification = (id: string) => {
+        setNotifications((prev) =>
+            prev.map((notification) => (notification.id === id ? { ...notification, open: false } : notification))
```

Comment:
this line can be: 

```
prev.map((notification) => ( { ...notification, open: notification.id !== id  })
```

Reply:
There might be benefits in having the shorter code variant but I considered going with my option for a couple of reasons: 
- have the code more readable as it is more obvious it changes the notification state only for the notification with the id selected while leaving the others unmodified
- have the open property modified only on the required object without introducing mutations on the irrelevant ones
- have a slight performance boost as we're not modifying all the items in the array


what do you think ?

Reply:
* I agree that it might read better but if we put `notification.id !== id` from the shorter variant into a `const isNotificationMarkedForClosing` it will improve the readability
* Actually with the shorter approach, nothing gets mutated, and new objects are created which is always a good practice, so it's safer for state changes
* map returns a new array with the modified elements so in either case, you are still doing O(N) operations where N is the size of the original array. It is true that you are not destructuring in some cases and you are destructuring only in one case and that will make your code O(N-1) + O(size of notification object which is destructured which is constant) which is still O(N) since we can drop the constants.